### ISSUE TITLE
Export AST-to-string API from `oxc-parser`?

### ISSUE THREAD

**Issue by leaysgur** [oxc-project/oxc#11495]
### Background

> Boshen:
> I wonder whether we should start forking and combining things to get things printed correct on the js side 😰 
> To complete the user story for oxc-parser-> transfer to js -> modify ast -> print
> https://discord.com/channels/1079625926024900739/1080712024683708466/1379710015539249273

I happened to be researching [`sveltejs/esrap`](https://github.com/sveltejs/esrap) for another task, so we thought it might be a good choice if it could be integrated easily.

### Prerequisite

`oxc-parser` currently produces:

- ESTree AST for JavaScript  
- TS-ESTree AST for TypeScript  

And it also supports JSX and several Stage 3 proposals.

### How could we implement this?

At first I planned to use `esrap` as an `optionalDependency` of `oxc-parser`.  

However, after trying it I found that:

- It does NOT support JSX
- It even fails to print simple TS code such as `let a: string | null = null`
- We would also need to attach comments to nodes manually

Another issue is that `sveltejs/esrap` expects the AST produced by `sveltejs/acorn-typescript`, which is similar to `@typescript-eslint/typescript-estree`(`oxc-parser` outputs), but it seems slightly [different](https://github.com/sveltejs/acorn-typescript/issues/7).

### Look for other implementations?

In short, the most robust way to turn a (TS-)ESTree AST into code today is Prettier. 😅

But if we choose Prettier, the main question is runtime performance? And it might be too much(too pretty) for this use-case.

If we do not choose Prettier, we could:

- Fork `esrap` and add support for more TS syntax and JSX
  - And also comment attaching
- Use `astring` or `estree-util-to-js` as `optionalDependencies` and wrap them
  - They have another package for comment attaching
  - (NOTE: I'm not yet tried actually)

Of course, the best option depends on how much effort we want to invest, but in any case Prettier will likely give more intuitive handling of comments than the alternatives.

However, if the comments do not have to be handled so strictly, performance seems to be better.

What do you think?

**Comment by jycouet** [comment]
Thank you for writing all this down, I'm going in circle for the last few days on all this :).
_I find it interesting that it's surfacing only now as the transform hook of `vite` expect to return some printed code. So you parse ✅, transform ✅ and you need to print it to return it somehow. 👀_

Coming from the svelte community, I stop at `esrap`, and want to improve a few not handled case _yet!_
I'll be happy to look at `let a: string | null = null`! (in addition to [the PR](https://github.com/sveltejs/esrap/pull/62) I already did yesterday)
Let's see how it's shaping 🚀

Side note, `oxc-parser` is removing comments, no ? Or I didn't see something ?

I'll be happy to help more!

**Comment by Boshen** [comment]
I think we can move forward with `esrap` if it covers jsx and conform to `@typescript-eslint/typescript-estree`.

> Side note, oxc-parser is removing comments, no ? Or I didn't see something ?

`oxc-parser` has not implemented any comment attachment logics (yet).

**Comment by leaysgur** [comment]
BTW, I plan to look into the following:

- [x] Read through `esrap`
  - I already know it does not yet support JSX
- [x] Check gaps between `acorn-typescript` & `@typescript-eslint/estree` (`oxc-parser`)
- [x] Check `astring` with `astravel` + `astring-jsx` + `astring-ts-generator` combo


---

As briefly mentioned on Discord, the ultimate(still-unseen!) future goal is likely:

- The AST on the JS side can be raw-transferred BACK to the Rust side  
- On the Rust side, we can turn that AST into code with `oxc_formatter`(or `oxc_codegen`?) with various options

...that's the ideal?

So, until that becomes possible, for the time being:

- For now, advise using a custom parser with Prettier  
  - A solid option that works today and supports both TS-ESTree and JSX  
  - Comment handling and quality are also ensured(not verified yet
- Meanwhile, provide some lightweight implementation ourselves?
  - How to implement depends on what I find in the investigation
  - Assume we'll compromise on comment handling, —best effort only  
  - Instead, it may be better in performance

**Comment by leaysgur** [comment]
I'll share my research results...

> if it covers jsx

No, `esrap` does not support printing JSX.

Also, there is no API to extend its behavior.

This means we need to fork it (or continuously send PRs?) to base on `esrap`.

> and conform to @typescript-eslint/typescript-estree.

Unfortunately, it does not.

`esrap` can print the AST generated by `@sveltejs/acorn-typescript`, but not all `TS`-prefixed node types are supported.

Their AST structure looks similar but is not the same.

For example:

- It does not produce `TSEnumBody`, `TSAsExpression`, `TSNonNullExpression`, etc
- It always produces `TSParenthesizedType`, `preserveParens: false` has no effect
- Leading comments do not update the `Program` node's `span.start`
- `module A.B.C {}` produces nested `TSModuleDeclaration` instead of a nested `TSQualifiedName`
- Our `add_fields` tweaks, like `typeAnnotation: null`, are not completely shown
- etc...

If you would like to see it in action, please try https://leaysgur.github.io/js-multi-ast-viewer/

**Comment by Boshen** [comment]
So there is no full solution for printing estree on the js side.

I added the "E-Help Wanted" label.

The core team does have the time and energy to maintain another codegen on the js side.

**Comment by leaysgur** [comment]
> no full solution for printing estree on the js side.

Let me elaborate on this a bit.

- The JS ESTree AST can be printed with `esrap`, `astring`, etc
  - For JSX, `astring` + `astring-jsx` can handle it
- For TS, `esrap` and `astring-ts-generator` can handle the `acorn-typescript` AST
  - But `acorn-typescript` does not fully support all TS syntaxes
- Finally, for the `typescript-estree` AST, only Prettier can print it correctly
- Otherwise, you need to extend `astring` via `generator` options and implement printer for JSX and TS-ESTree
  - `esrap` may also become extendable in near [future](https://github.com/sveltejs/esrap/issues/35)

**Comment by overlookmotel** [comment]
> As briefly mentioned on Discord, the ultimate(still-unseen!) future goal is likely:
> 
> * The AST on the JS side can be raw-transferred BACK to the Rust side
> * On the Rust side, we can turn that AST into code with `oxc_formatter`(or `oxc_codegen`?) with various options
> 
> ...that's the ideal?

Just to say, yes, I think this is our ideal end goal. But we don't expect to get there any time soon.

**Comment by Rich-Harris** [comment]
Hey, Rich from the Svelte team here. Our plan for `esrap` is to make it pluggable (https://github.com/sveltejs/esrap/issues/35) in such a way that it can print any AST that uses the `{ type: string, ... }` shape, be it JS, TS, JSX, CSS or whatever else. It's been on our wishlist for a while but if being able to add things like JSX support would be valuable to you, I'm happy to bump it up the priority list

**Comment by leaysgur** [comment]
Thanks for stopping by~! 🙏🏻 

As for the full set we are looking for, it seems not easy since we need to support not only JSX but also TS-ESTree (from `typescript-estree`, not from `acorn-typescript`).

However, it is very good news that `esrap` will be extensible!

**Comment by Rich-Harris** [comment]
There should be no problem supporting TS-ESTree

**Comment by leaysgur** [comment]
Ah, sorry! once `esrap` is extensible, there's no problem; the only issue is our resource limitations. 😅

**Comment by Rich-Harris** [comment]
We just published esrap 2, with TSESTree and experimental JSX support: https://github.com/sveltejs/esrap

**Comment by leaysgur** [comment]
Thanks for letting us know~!

---

FYI: I did a little testing at hand, and it seems that a few more works are needed to fully support TSEStree produced by `@typescript-eslint/typescript-estree` and `oxc-parser`.

```
👀 [print-throws] Cannot read properties of undefined (reading 'loc')
👀 [print-throws] Cannot read properties of undefined (reading 'map')
👀 [print-throws] Cannot read properties of undefined (reading 'typeAnnotation')
👀 [print-throws] Not implemented: AccessorProperty
👀 [print-throws] Not implemented: TSAbstractMethodDefinition (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSAbstractPropertyDefinition (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSBigIntKeyword (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSCallSignatureDeclaration (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSClassImplements (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSConstructSignatureDeclaration (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSConstructorType (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSExportAssignment (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSImportEqualsDeclaration (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSInferType (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSInstantiationExpression (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSInterfaceHeritage (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSIntrinsicKeyword (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSMappedType (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSNamespaceExportDeclaration (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSObjectKeyword (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSOptionalType (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSParameterProperty (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSParenthesizedType (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSRestType (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSTemplateLiteralType (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSThisType (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSTypeAssertion (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSTypeOperator (consider using 'esrap/languages/ts')
👀 [print-throws] Not implemented: TSTypePredicate (consider using 'esrap/languages/ts')
```

I'd like to submit a PR if I can find the time, but in the meantime...

<details>
<summary>Output with this code</summary>

```js
import { readFile } from "node:fs/promises";
import { glob } from "tinyglobby";
import { parseSync } from "oxc-parser";
import { print } from "esrap";
import ts from "esrap/languages/ts";
import tsx from "esrap/languages/tsx";

const PRETTIER_DIR = "../../Codes/prettier";
const PRETTIER_TESTS_DIR = `${PRETTIER_DIR}/tests/format`;

const printThrows = new Set();
for await (const file of getPrettierTestFiles()) {
  const code = await readFile(`${PRETTIER_TESTS_DIR}/${file}`, "utf8");

  let parseRet1;
  try {
    parseRet1 = parseSync(file, code, {
      experimentalRawTransfer: true,
    });
  } catch (err) {
    // console.error(file);
    // console.error("💥", "[parseSync-throws]", err);
    // console.error();
    continue;
  }
  if (parseRet1.errors.length !== 0) {
    // console.error(file);
    // console.error("💥", "[parseSync-errors]", parseRet1.errors);
    // console.error();
    continue;
  }

  let printRet;
  try {
    const visitors = file.endsWith("sx")
      ? tsx({ comments: parseRet1.comments })
      : ts({ comments: parseRet1.comments });
    printRet = print(parseRet1.program, visitors, {});
  } catch (err) {
    console.error(file);
    console.error("👀", "[print-throws]", err);
    console.error();

    printThrows.add(err.message);
    continue;
  }

  let parseRet2;
  try {
    parseRet2 = parseSync(file, printRet.code, {
      experimentalRawTransfer: true,
    });
  } catch (err) {
    console.error(file);
    console.error("💥", "[parseSync2-throws]", err);
    console.error();
    continue;
  }

  // TODO: Check if parseRet1 === parseRet2
}

console.log("✨", "All files processed.");
for (const printThrow of Array.from(printThrows).sort()) {
  console.error("👀", `[print-throws] ${printThrow}`);
}

// ---

async function* getPrettierTestFiles() {
  for (const file of await glob(["js", "jsx", "typescript"], {
    cwd: PRETTIER_TESTS_DIR,
    ignore: ["**/*format.test.js", "**/*.snap"],
  })) {
    if (file.includes("babel")) continue;
    yield file;
  }
}
```

</details>

**Comment by Rich-Harris** [comment]
That's very helpful, thanks

**Comment by manuel3108** [comment]
> FYI: I did a little testing at hand, and it seems that a few more works are needed to fully support TSEStree produced by `@typescript-eslint/typescript-estree` and `oxc-parser`.

I have started working on the list you provided, and I am now down to 14 fails instead of 29 as shown above. Seems to be working without too much hustle overall.

In the meantime, together with @jycouet we found out that comments are handled differently as well, and that `esrap` is currently not at all printing comments. @jycouet identified that this is again due to a different handling of comments between `acorn` and `oxc-parser`. In acorn, comments have a loc `property` giving you information about the line and the start and end it's in. In `oxc-parser` the information for the line seems to be omitted. Which currently leads us to [this massive pr](https://github.com/manuel3108/esrap/pull/3) trying to guess on where the comments should go (and guessing obviously only works in some circumstances).

Are we missing something? Has there already been previous work on the `oxc` side to determine where those comments should actually be printed, so that we could have a look? Just joined the discord, in case you feel that we should continue the discussion there

**Comment by leaysgur** [comment]
Thank you for your efforts!

#10307 is the exact issue for adding `range` and `loc` to the OXC's ESTree AST.

Certainly that is planned, but it will take some time to get there. 🏃🏻

**Comment by manuel3108** [comment]
Perfect, thank you. We will leave comments untouched for now, then!

**Comment by mizulu** [comment]
I put this in the oxc discord. few days ago
but didn't get much feedback, I am not sure if it esrap or oxc fault
so I am just going to put this here, in case there are any efforts to make esrap and oxc compatible 



for the esrap to be able to print this I had to patch it.  ( I have added comment in each line I had to change ) 
```ts
export type FilterFunction = (context: MutationContext) => FilterResultEnum;
```
The patching 
```ts
TSFunctionType(node, context) {
            if (node.typeParameters) context.visit(node.typeParameters);

            context.write('(');

            sequence(
                context,
                // @ts-expect-error `acorn-typescript` and `@typescript-eslint/types` have slightly different type definitions
                
                node.parameters 
                   || node.params, // added params for oxc

                // @ts-expect-error `acorn-typescript` and `@typescript-eslint/types` have slightly different type definitions
                node?.returnType?.typeAnnotation?.start ?? // oxc: fix
                node?.typeAnnotation?.typeAnnotation.loc?.start ?? null,
                false
            );

            context.write(') => ');

            // @ts-expect-error `acorn-typescript` and `@typescript-eslint/types` have slightly different type definitions
       context.visit(
               node?.typeAnnotation?.typeAnnotation
                ?? node.returnType.typeAnnotation // oxc: fix
            ); 
            
        },
```

there is 2 differences between what esrap was expecting and what the parser produced. 

from
```
node?.typeAnnotation?.typeAnnotation.loc?.start` 
```
to 
```
node?.returnType?.typeAnnotation?.start
```

from 
```
node.parameters
```

to 
```
node.params
```

**Comment by mizulu** [comment]
Can someone also explain to me how the oxc playground is able to do "codegen" 
and why do we need a JS implementation and it is not delegated to the wasm the same way we can use the parser ?

**Comment by leaysgur** [comment]
> I put this in the oxc discord. few days ago but didn't get much feedback,

Sorry, I have been in bed with a fever recently, so I didn't pick it up...

Now, I try to answer your questions. 💪🏻 

---

> for the esrap to be able to print this I had to patch it.

This is exactly what we were talking about in this issue.

Currently, `esrap` expects the AST from `@sveltejs/acorn-typescript`.

And AST produced by `oxc-parser` is not directly compatible. Our output is the same one with `@typescript-eslint/(parser|typescript-estree)`'s output.

As mentioned above, there is an ongoing [PR](https://github.com/sveltejs/esrap/pull/70) to reduce these gaps in `esrap` repo. 👀 

I'm not sure when it will be merged, but I think your patches are covered as well. (You can check the PR and follow up?)

> how the oxc playground is able to do "codegen"

In the playground, AST is converted to string in Rust side directly with our Rust AST.

https://github.com/oxc-project/oxc/blob/main/napi/playground/src/lib.rs#L255-L256

> and why do we need a JS implementation
> and it is not delegated to the wasm the same way we can use the parser ?

Because Rust side AST and JS side AST(ESTree) is not a same shape.

If your usecase is written in Rust, I think you can use the [`oxc_codegen`](https://crates.io/crates/oxc_codegen) crate.

I don't think there is much demand to use OXC's Rust ASTs as is on the JS side, so `oxc-parser` outputs (TS-)ESTree AST.

Then you may want to know how to convert (TS-)ESTree AST to string, which is... the purpose of this issue.

**Comment by mizulu** [comment]
> > I put this in the oxc discord. few days ago but didn't get much feedback,
> 
> Sorry, I have been in bed with a fever recently, so I didn't pick it up...
> 
> Now, I try to answer your questions. 💪🏻
> 
> > for the esrap to be able to print this I had to patch it.
> 
> This is exactly what we were talking about in this issue.
> 
> Currently, `esrap` expects the AST from `@sveltejs/acorn-typescript`.
> 
> And AST produced by `oxc-parser` is not directly compatible. Our output is the same one with `@typescript-eslint/(parser|typescript-estree)`'s output.
> 

Thank you for the response and clarification.  

> As mentioned above, there is an ongoing [PR](https://github.com/sveltejs/esrap/pull/70) to reduce these gaps in `esrap` repo. 👀

👍 

> 
> I'm not sure when it will be merged, but I think your patches are covered as well. (You can check the PR and follow up?)
> 
> > how the oxc playground is able to do "codegen"
> 
> In the playground, AST is converted to string in Rust side directly with our Rust AST.
> 
> https://github.com/oxc-project/oxc/blob/main/napi/playground/src/lib.rs#L255-L256
> 


> > and why do we need a JS implementation
> > and it is not delegated to the wasm the same way we can use the parser ?
> 
> Because Rust side AST and JS side AST(ESTree) is not a same shape.
> 
> If your usecase is written in Rust, I think you can use the [`oxc_codegen`](https://crates.io/crates/oxc_codegen) crate.
> 
> I don't think there is much demand to use OXC's Rust ASTs as is on the JS side, so `oxc-parser` outputs (TS-)ESTree AST.
> 
> Then you may want to know how to convert (TS-)ESTree AST to string, which is... the purpose of this issue.

My use case was actually fairly simple, I just had to remove some nodes (exports / imports_  from the ast, instead of using regexp 
I did not have any particular ast format need.  so even if I had access to the rust ast in js and the codgen that would have been fine for my usecase.

**Comment by manuel3108** [comment]
I wasn't pushing for the mentioned PR to be merged, because I did not see any meaningful benefit. That's because comments still wont work, as the expected properties are not present. I thought it was better to fail with an error, rather than swallowing the comments

**Comment by manuel3108** [comment]
We just published `esrap@2.1.1` that should fix everything reported here (and way more).

@leaysgur Thanks for providing your reproduction code, that massively helped finding and fixing all the missing implementations

**Comment by leaysgur** [comment]
Thank you for letting us know, and for working on this! Amazing work~! 👏🏻 

Let's close this issue as `esrap` supports printing the (TS-)ESTree AST parsed by `oxc-parser`.

**Comment by TheAlexLichter** [comment]
@leaysgur should we add an official example to the docs?

**Comment by jycouet** [comment]
I'll be happy to do it ☺️

Note that to be really complete, we need `loc` to print comments (#10307)

**Comment by leaysgur** [comment]
Let’s do it. 👍🏻

@TheAlexLichter Would you please create an issue for this in https://github.com/oxc-project/oxc-project.github.io/issues and assign @jycouet ?

Also, share rough guidance if you have recommendations?
Like:
- should add a few sentences under guides/parser page
- or create another page for ecosystem tools
- or etc

Thank you both in advance! 🙏🏻
